akka-actor-typed/src/main/scala/akka/actor/typed/StashBuffer.scala

```diff
@@ -0,0 +1,332 @@
+   *
+   * @param message the message to buffer
+   * @return this message buffer
+   */
+  def append(message: T): ImmutableStashBuffer[T] = {
+    if (message == null) throw new NullPointerException
+    if (isFull)
+      throw new StashOverflowException(s"Couldn't add [${message.getClass.getName}] " +
+        s"because stash with capacity [$capacity] is full")
+
+    new ImmutableStashBuffer(capacity, buffer :+ message)
+  }
+
+  /**
+   * Add one element to the end of the message buffer. Note that this class is
+   * immutable so the returned instance contains the added message.
+   *
+   * [[StashOverflowException]] is thrown if the buffer [[MutableStashBuffer#isFull]].
+   *
+   * @param message the message to buffer
+   * @return this message buffer
+   */
+  def :+(message: T): ImmutableStashBuffer[T] =
+    append(message)
+
+  /**
+   * Remove the first element of the message buffer. Note that this class is
+   * immutable so the head element is removed in the returned instance.
+   *
+   * @throws `NoSuchElementException` if the buffer is empty
+   */
+  def dropHead(): ImmutableStashBuffer[T] =
+    if (buffer.nonEmpty) new ImmutableStashBuffer(capacity, buffer.tail)
+    else throw new NoSuchElementException("head of empty buffer")
+
+  /**
+   * Return the first element of the message buffer.
+   *
+   * @return the first element or throws `NoSuchElementException` if the buffer is empty
+   * @throws `NoSuchElementException` if the buffer is empty
+   */
+  def head: T =
+    if (buffer.nonEmpty) buffer.head
+    else throw new NoSuchElementException("head of empty buffer")
+
+  /**
+   * Iterate over all elements of the buffer and apply a function to each element.
+   *
+   * @param f the function to apply to each element
+   */
+  def foreach(f: T ⇒ Unit): Unit =
+    buffer.foreach(f)
+
+  /**
+   * Java API
+   *
+   * Iterate over all elements of the buffer and apply a function to each element.
+   *
+   * @param f the function to apply to each element
+   */
+  def forEach(f: Consumer[T]): Unit = foreach(f.accept)
+
+  /**
+   * Scala API: Process all stashed messages with the `behavior` and the returned
+   * [[Behavior]] from each processed message.
+   */
+  def unstashAll(ctx: scaladsl.ActorContext[T], behavior: Behavior[T]): Behavior[T] =
+    internalUnstashAll(ctx.asInstanceOf[ActorContext[T]], behavior)
+
+  /**
+   * Java API: Process all stashed messages with the `behavior` and the returned
+   * [[Behavior]] from each processed message.
+   */
+  def unstashAll(ctx: javadsl.ActorContext[T], behavior: Behavior[T]): Behavior[T] =
+    internalUnstashAll(ctx.asInstanceOf[ActorContext[T]], behavior)
+
+  /**
+   * Process all stashed messages with the `behavior` and the returned
+   * [[Behavior]] from each processed message.
+   */
+  private def internalUnstashAll(ctx: ActorContext[T], behavior: Behavior[T]): Behavior[T] = {
+    buffer.foldLeft(behavior) { (b, msg) ⇒
```

Comment:
@johanandren perhaps you know, if `undefer` is needed here? You have it in https://github.com/akka/akka/pull/24302/files#diff-952d6f5ecc93ffa8f232d6e149bd17c1R240

btw, the stashing in the guardian should also use this

Reply:
I think so, if behavior is `deferred(stop)` then `isAlive(behavior)` will still be true and there isn't really a way around that.